plugins/inputs/sqlserver/azuresqlqueries.go

```diff
@@ -679,6 +679,138 @@ SELECT
 FROM sys.dm_os_schedulers AS s
 `
 
+const sqlAzureDBPartQueryPeriod = `
+DECLARE @QueryData NVARCHAR(MAX) = ?1; --input parameter
+
+IF SERVERPROPERTY('EngineEdition') <> 5 BEGIN /*not Azure SQL DB*/
+	DECLARE @ErrorMessage AS nvarchar(500) = 'Telegraf - Connection string Server:'+ @@SERVERNAME + ',Database:' + DB_NAME() +' is not an Azure SQL DB. Check the database_type parameter in the telegraf configuration.';
+	RAISERROR (@ErrorMessage,11,1)
+	RETURN
+END;
+
+DECLARE @lastIntervalEndTimestamp BIGINT = 0;
+
+/* Get the time when the query was previously called */
+IF @QueryData != '' AND @QueryData IS NOT NULL
+	SET @lastIntervalEndTimestamp = CAST(@QueryData AS BIGINT);
+
+DECLARE @secondsInDay INT = 86400;
+
+/* Get query aggregation interval from query store */
+DECLARE @timeGrain BIGINT = (SELECT interval_length_minutes * 60 FROM sys.database_query_store_options WITH (NOLOCK)); 
+
+/* Get current time and round it down to a multiple of @timeGrain */
+DECLARE @currDate datetimeoffset =  SYSDATETIMEOFFSET();
+DECLARE @currIntervalEndTimestamp BIGINT = DATEDIFF_BIG(second, 0, @currDate) / @timeGrain * @timeGrain;
+
+/* If we are still in the same @timeGrain interval with when the query was previously called, quit and do not report any data */
+IF @lastIntervalEndTimestamp = @currIntervalEndTimestamp
+	RETURN;
+
+/* Calculate the start time for returning data from query store:
+If the query was never called before, start time is end time minus @timeGrain */
+DECLARE @currIntervalStartTimestamp BIGINT = @currIntervalEndTimestamp - @timeGrain;
+
+/* If the query was called before, start time is when the query was previously called, so that we return all the records since the previous call */
+IF @lastIntervalEndTimestamp != 0
+	SET @currIntervalStartTimestamp = @lastIntervalEndTimestamp;
+
+/* Convert start and end times to datetime */
+DECLARE @currIntervalEndDate datetimeoffset = DATEADD(second, @currIntervalEndTimestamp % @secondsInDay, DATEADD(day, @currIntervalEndTimestamp / @secondsInDay, 0));
+DECLARE @currIntervalStartDate datetimeoffset = DATEADD(second, @currIntervalStartTimestamp % @secondsInDay, DATEADD(day, @currIntervalStartTimestamp / @secondsInDay, 0));
```

Comment:
@dimitri-furman, what about using **AT TIME ZONE 'UTC'** everywhere we need to compare or output datetimeoffsets? It may help to avoid issues related to DST. And it looks like DST is not applicable to Azure SQL DB, it always uses UTC.

Reply:
It may work functionally for time zone issues on MI and SQL VM. But there could be perf impact in Query Store collection queries because we would need to use AT TIME ZONE on the column side of the predicate, and that will prevent index seeks.

The change would do nothing to solve potential problems when Query Store interval duration is changed.

More generally, I don't see the value in back and forth conversions between bigint and datetimeoffset, they seem to just add unnecessary complexity. Am I overlooking the goal there? It just seems overall easier to cache datetimeoffset if that's what we actually use in the queries to filter.

Reply:
@dimitri-furman, the interval calculation logic was changed and now it doesn't use bigint.

Reply:
Thanks for making this change, looks good.

I haven't looked at Managed Instance and SQL Server queries, but for all recent changes made for SQL DB Query Store collection, similar changes should be made for MI and SQL Server too.